Add Deep Learning Blog 2 (#61)
mlflow/mlflow-website: website/blog/2024-04-26-deep-learning-part-2/index.md

most people simply accept them without reading them. However, reports have indicated over time that occasionally, some suspiciously unfair
terms are embedded within them.

Addressing unfair clauses in Terms of Service (TOS) agreements through machine learning (ML) is particularly relevant due to the pressing
need for transparency and fairness in legal agreements that affect consumers. Consider the following clause from an example TOS
agreement: **"We may revise these Terms from time to time. The changes will not be retroactive, and the most current version of the Terms, which will always..."**
This clause stipulates that the service provider may suspend or terminate the service at any time for any reason,
with or without notice. Most people would consider this to be quite unfair.

While this sentence is buried quite deeply within a fairly dense document, an ML algorithm is not burdened by the exhaustion that a human
would have for combing through the text and identifying clauses that might seem a bit unfair. By automatically identifying potentially

> unfair clauses, transformers can help protect consumers from exploitative practices, ensure greater compliance with legal standards,

Reviewer: nit: `transformers can help protect customers...` sounds a bit weird protect consumers: sounds a bit weird as transformers us a framework not a model, like we don't say `python can help....`. Can we rephrase it to "the ML model can help ..."?

and foster trust between service providers and users.

> A base transformer model, without specialized fine-tuning, faces several challenges in accurately identifying unfair Terms of Service clauses.

Reviewer: ```suggestion
A base pretrained model, without specialized fine-tuning, faces several challenges in accurately identifying unfair Terms of Service clauses.
```
nit: transformer is an architecture name

contextual meanings that determine the fairness of contractual terms, making it less effective for this specialized task.

Using prompt engineering to address the identification of unfair Terms of Service clauses with a closed-source Large language model
can be prohibitively expensive. This approach requires extensive trial and error to refine prompts without the ability to tweak
the underlying model mechanics. Each iteration can consume significant computational resources , especially when using
[few-shot prompting](https://www.promptingguide.ai/techniques/fewshot), leading to escalating costs without guaranteeing a corresponding
increase in accuracy or effectiveness.

In this context, the use of the **RoBERTa-base** model is particularly effective, provided that it is fine-tuned. This model is robust
enough to handle complex tasks like discerning embedded instructions within texts, yet it is sufficiently compact to be fine-tuned
on modest hardware, such as an Nvidia T4 GPU.

### What is PEFT?

[Parameter-Efficient Fine-Tuning (PEFT)](https://huggingface.co/docs/peft/en/index) approaches are advantageous as they involve

> keeping the bulk of the pre-trained model parameters fixed while only training a few additional layers. This methodology not only

Reviewer: ```suggestion
keeping the bulk of the pre-trained model parameters fixed while only training a fairy small percentage of parameters.
```
nit: Not all PEFT methods uses additional layers.

        """
        model = AutoModelForSequenceClassification.from_pretrained(
            pretrained_model_name_or_path=pretrained_model,
            num_labels=num_classes
        )
        peft_config = LoraConfig(
            task_type=TaskType.SEQ_CLS,
            inference_mode=False,
            r=r,
            lora_alpha=lora_alpha,
            lora_dropout=lora_dropout
        )
        return get_peft_model(model, peft_config)

    def forward(self, input_ids: List[int], attention_mask: List[int], label: List[int]):
        """Calculate the loss by passing inputs to the model and comparing against ground truth labels.

        Args:
            input_ids: List of token indices to be fed to the model.
            attention_mask: List to indicate to the model which tokens should be attended to, and which should not.
            label: List of ground truth labels associated with the input data.

        Returns:
            torch.Tensor: The computed loss from the model as a tensor.
        """
        return self.model(
            input_ids=input_ids,
            attention_mask=attention_mask,
            labels=label
        )
```

> Additional references for the full implementation can be [seen within the companion repository here](https://github.com/puneet-jain159/deeplearning_with_mlfow/blob/master/custom_module/fine_tune_clsify_head.py)

Reviewer: nit: not related to the blog content, but the linked code has a bit incorrect Python syntax
https://github.com/puneet-jain159/deeplearning_with_mlfow/blob/master/custom_module/fine_tune_clsify_head.py#L60-L100

There are two `return` in the `predict` function where anything below L77 is simply discarded.

Reviewer 2: Hi @B-Step62  
Corrected the code.  and tested the updated code with an end 2 end run. 

Link : https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/4093282202515276

as demonstrated below.

### Configuring Pytorch Trainer Callback with Early stopping

The example below shows the configuration of the `Trainer` object within `Lightning` to leverage early stopping to prevent overfitting. Once configured, the
training is executed by calling `fit` on the `Trainer` object. By providing the `EarlyStopping` callback, in conjunction with MLflow's autologging, the
appropriate number of epochs will be used, logged, and tracked without any additional effort.

```python
from lightning import Trainer
from lightning.pytorch.callbacks import EarlyStopping

# Run the training loop.
trainer = Trainer(
   callbacks=[
       EarlyStopping(
           monitor="Val_F1_Score",
           min_delta=train_config.min_delta,
           patience=train_config.patience,
           verbose=True,
           mode="max",
       )
   ],
   default_root_dir=train_config.model_checkpoint_dir,
   fast_dev_run=bool(train_config.debug_mode_sample),
   max_epochs=train_config.max_epochs,
   max_time=train_config.max_time,
   precision="32-true"
)

> trainer.fit(model=nlp_model, datamodule=datamodule)

Reviewer: This `nlp_model` is not defined above and suddenly appears. I think it's crucial part as it contains pretrained RoBERTa model name and LoRA configurations. Can we add that code snippet?

loop. Integrating the MLflow callback into the PyTorch Lightning Trainer allows for iterative logging of loss metrics at
configurable iterations, enabling an easily debuggable evaluation of the training performance, ensuring that stopping criteria
can be enforced at the appropriate time to prevent overfitting.

### Evaluating epoch checkpoints of Fine Tuned Models with MLflow

With your training process meticulously tracked and logged by MLflow, you have the flexibility to retrieve and test your model at
any arbitrary checkpoint. To do this, you can use the mlflow.pytorch.load_model() API to load the model from a specific run
and use the `predict()` method for evaluation.

In the example below, we will load the model checkpoint from the 3rd epoch and use the `Lightning` train module to generate predictions based on the
checkpoint state of the saved training epoch.

```python
import mlflow

mlflow.pytorch.autolog(disable = True)

run_id = '<Add the run ID>'
model = mlflow.pytorch.load_model(f"runs:/{run_id}/model")

model = mlflow.pytorch.load_checkpoint(TransformerModule, run_id, 3)

> train_module.predict(model, dataloaders = DataLoader(ds))

Reviewer: Do we need to do this instead of `model.predict()?`

Reviewer 2: Yes. because the loaded object is a Pytorch object (not a pyfunc object) which does not have a predict method. 

`type(model)`
> custom_module.fine_tune_clsify_head.TransformerModule

```

## Summary

The integration of MLflow into the finetuning process of pre-trained language models, particularly for applications like custom
named entity recognition, text classification and instruction-following represents a significant advancement in managing and
optimizing deep learning workflows. Leveraging the autologging and tracking capabilities of MLflow in these workstreams not only
enhances the reproducibility and efficiency of model development, but also fosters a collaborative environment where insights
and improvements can be easily shared and implemented.

As we continue to push the boundaries of what these models can achieve, tools like MLflow will be instrumental in harnessing their full potential.

If you're interested in seeing the full example in its entirety, feel free to [see the full example implementation](https://github.com/puneet-jain159/deeplearning_with_mlfow)

### Check out the code

> The code we provide will delve into additional aspects such as training from a checkpoint, integrating MLflow and TensorBoard, and utilizing Pyfunc for model wrapping, among others. These resources are specifically tailored for implementation on [Databricks Community Edition](https://mlflow.org/blog/databricks-ce).

Reviewer: nit: Can we add link to the repo here? (Ideally a notebook that users can import to DCE)
